### ISSUE TITLE
Custom CSS checkbox disappearing with acceptance validation.

### ISSUE THREAD

**Issue by markcatley** [DavyJonesLocker/client_side_validations#775]
### Steps to reproduce*

Click on the checkbox to check it, then uncheck it.

NOTE: Other client-side validations work correctly on this form. I would suspect this is to do with the way I've done a CSS checkbox. Although I am not sure what I should do it to fix it on my end. It seems like it might be a common problem, but I couldn't find anything in the readme or on the wiki. My apologies if there is somewhere else I should have looked.

### Expected behavior*
The checkbox should check and then uncheck.

### Actual behavior*
The checkbox checks and then disappears

### System configuration*
**Rails version**: 6.0.1

**Ruby version**: 2.6.5p114

**Client Side Validations version**: 16.0.3

### Code snippet from your model of the validations*
```
    validates :first_name, presence: true
    validates :last_name, presence: true
    validates :person_email, confirmation: true, presence: true
    validates :accept_and_read_terms_of_use, acceptance: true
```
### The whole form code from your template*
```
  <%= form_for @account, url: public_accounts_path, validate: true, html: { class: 'grid-y', novalidate: true } do |f| %>
    <% if @account.errors.any? %>
      <div class="error-explanation">
        <h2 class="h3"><%= pluralize(@account.errors.count, 'error') %> prevented the account from being created:</h2>

        <ul>
        <% @account.errors.full_messages.each do |msg| %>
          <li><%= msg %></li>
        <% end %>
        </ul>
      </div>
    <% end %>
    <div class="grid-x grid-margin-x align-middle">
      <div class="cell small-12 medium-6">
        <%= f.text_field :first_name, placeholder: 'Last name', class: 'cell ii-person', required: true %>
        <%= f.text_field :last_name, placeholder: 'First name', class: 'cell ii-person', required: true %>
        <%= f.email_field :person_email, placeholder: 'Email', class: 'cell ii-mail', required: true %>
        <%= f.email_field :person_email_confirmation, placeholder: 'Confirm email', class: 'cell ii-mail', required: true %>
        <%= f.telephone_field :person_mobile_phone, placeholder: 'Phone number', class: 'cell ii-phone', required: true %>
      </div>
      <div class="cell small-12 medium-6">
        <%= f.label :accept_and_read_terms_of_use, class: 'cell' do %>
          <div class="grid-x grid-margin-x">
            <div class="cell shrink grid-x align-middle">
              <%= f.check_box :accept_and_read_terms_of_use %>
              <%= f.label :accept_and_read_terms_of_use do %>
              <% end %>
            </div>
            <div class="cell auto">
              I have read and I agree to the terms and conditions
            </div>
          </div>
        <% end %>

        <div class="grid-x grid-margin-x">
          <div class="cell small-12 large-6">
            <%= f.submit 'Sign up', class: 'button primary expanded form-submit-2' %>
          </div>
          <div class="cell small-12 large-6">
            <%= link_to 'Login', '', class: 'button hollow expanded cell small-12 large-6 form-submit-2' -%>
          </div>
        </div>
      </div>
    </div>
  <% end %>
```
### The resulting HTML*
```
  <form class="grid-y" id="new_public_new_account" novalidate="novalidate" data-client-side-validations="{&quot;html_settings&quot;:{&quot;type&quot;:&quot;ActionView::Helpers::FormBuilder&quot;,&quot;input_tag&quot;:&quot;\u003cdiv class=\&quot;field_with_errors\&quot;\u003e\u003cspan id=\&quot;input_tag\&quot; /\u003e\u003clabel for=\&quot;\&quot; class=\&quot;message form-field-error-message\&quot;\u003e\u003c/label\u003e\u003c/div\u003e&quot;,&quot;label_tag&quot;:&quot;\u003cdiv class=\&quot;field_with_errors\&quot;\u003e\u003clabel id=\&quot;label_tag\&quot; /\u003e\u003c/div\u003e&quot;},&quot;number_format&quot;:{&quot;separator&quot;:&quot;.&quot;,&quot;delimiter&quot;:&quot;,&quot;},&quot;validators&quot;:{&quot;public_new_account[first_name]&quot;:{&quot;presence&quot;:[{&quot;message&quot;:&quot;can&#39;t be blank&quot;}]},&quot;public_new_account[last_name]&quot;:{&quot;presence&quot;:[{&quot;message&quot;:&quot;can&#39;t be blank&quot;}]},&quot;public_new_account[person_email]&quot;:{&quot;confirmation&quot;:[{&quot;message&quot;:&quot;doesn&#39;t match Person email&quot;,&quot;case_sensitive&quot;:true}],&quot;presence&quot;:[{&quot;message&quot;:&quot;can&#39;t be blank&quot;}]},&quot;public_new_account[accept_and_read_terms_of_use]&quot;:{&quot;acceptance&quot;:[{&quot;message&quot;:&quot;must be accepted&quot;,&quot;accept&quot;:[&quot;1&quot;,true]}]}}}" action="/public/accounts" accept-charset="UTF-8" method="post"><input type="hidden" name="authenticity_token" value="uFoKuCxwqZjUIpGbFXmwEvwl/72/fq3+Q/PGbr4Iw9zXH8kNP4hOMPPwpufi7lPGhzss+8DQvaHed6T3nqh/fg==" />
    <div class="grid-x grid-margin-x align-middle">
      <div class="cell small-12 medium-6">
        <input placeholder="Last name" class="cell ii-person" required="required" type="text" name="public_new_account[first_name]" id="public_new_account_first_name" />
        <input placeholder="First name" class="cell ii-person" required="required" type="text" name="public_new_account[last_name]" id="public_new_account_last_name" />
        <input placeholder="Email" class="cell ii-mail" required="required" type="email" name="public_new_account[person_email]" id="public_new_account_person_email" />
        <input placeholder="Confirm email" class="cell ii-mail" required="required" type="email" name="public_new_account[person_email_confirmation]" id="public_new_account_person_email_confirmation" />
        <input placeholder="Phone number" class="cell ii-phone" required="required" type="tel" name="public_new_account[person_mobile_phone]" id="public_new_account_person_mobile_phone" />
      </div>
      <div class="cell small-12 medium-6">
        <label class="cell" for="public_new_account_accept_and_read_terms_of_use">
          <div class="grid-x grid-margin-x">
            <div class="cell shrink grid-x align-middle">
              <input name="public_new_account[accept_and_read_terms_of_use]" type="hidden" value="0" /><input type="checkbox" value="1" name="public_new_account[accept_and_read_terms_of_use]" id="public_new_account_accept_and_read_terms_of_use" />
              <label for="public_new_account_accept_and_read_terms_of_use">
</label>            </div>
            <div class="cell auto">
              I have read and I agree to the
              <a href="">Syndex Account Holder Rules</a> and
              <a href="">Privacy Policy</a>            </div>
          </div>
</label>
        <div class="grid-x grid-margin-x">
          <div class="cell small-12 large-6">
            <input type="submit" name="commit" value="Sign up" class="button primary expanded form-submit-2" data-disable-with="Sign up" />
          </div>
          <div class="cell small-12 large-6">
            <a class="button hollow expanded cell small-12 large-6 form-submit-2" href="">Login</a>          </div>
        </div>
      </div>
    </div>
</form>
```
### Browser's development console output*
There was an error, unfortunately I don't understand it. Here it is:
```
jquery.js:5436 Uncaught DOMException: Failed to execute 'replaceChild' on 'Node': The new child element contains the parent.
    at HTMLLabelElement.<anonymous> (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:12871:20)
    at domManip (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:12627:20)
    at jQuery.fn.init.replaceWith (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:12864:14)
    at Object.add (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:1941:51)
    at Object.addError (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:1874:111)
    at http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:1846:47
    at Object.fail (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:1788:16)
    at HTMLInputElement.elementValidateFailClientSideValidations (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:1845:51)
    at HTMLInputElement.dispatch (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:12023:98)
    at HTMLInputElement.elemData.handle (http://localhost:3000/packs/js/public-46cb04eaba9be0486ddf.js:11843:109)
```
### Additional JavaScript Libraries*
_If your issue depends on other JavaScript libraries, please list them here. E.g: *Bootstrap Modal v3.3.7, jQuery UI Datepicker 1.12.4*._

### Repository demostrating the issue
Debugging CSV issues is a time consuming task. If you want to speed up things, please
provide a link to a repository showing the issue.

I am more than happy to do this if needed but it will take me a bit to extract it. Please let me know if this would be helpful. I suspect that someone with experience will be able to point out my mistake quickly.

---

**\*** Failure to include this requirement may result in the issue being closed.

**Comment by markcatley** [comment]
FYI, I've figured this out. I needed to do a custom add/remove method. Here's what I did:
```javascript
window.ClientSideValidations.formBuilders['ActionView::Helpers::FormBuilder'] = {
  add: function(element, settings, message) {
    const form = $(element[0].form)

    if (element.data('valid') !== false && (form.find("label.message[for='" + (element.attr('id')) + "']")[0] == null)) {
      const inputErrorField = $(settings.input_tag)
      const labelErrorField = $(settings.label_tag)
      if (element.attr('autofocus')) {
        element.attr('autofocus', false)
      }

      if (element.attr('type') != 'checkbox') {
        const label = form.find("label[for='" + (element.attr('id')) + "']:not(.message)")[0]

        element.before(inputErrorField)
        inputErrorField.find('span#input_tag').replaceWith(element)
        inputErrorField.find('label.message').attr('for', element.attr('id'))
        labelErrorField.find('label.message').attr('for', element.attr('id'))
        labelErrorField.insertAfter(label)
        labelErrorField.find('label#label_tag').replaceWith(label)
      } else {
        const container = element.parents('label');
        container.before(inputErrorField)
        inputErrorField.find('span#input_tag').replaceWith(container)
        inputErrorField.find('label.message').attr('for', element.attr('id'))
        labelErrorField.find('label.message').attr('for', element.attr('id'))
        labelErrorField.insertAfter(null)
        labelErrorField.find('label#label_tag').replaceWith(null)
      }
    }
    form.find("label.message[for='" + (element.attr('id')) + "']").text(message)
  },

  remove: function(element, settings) {
    const form = $(element[0].form)
    const errorFieldClass = $(settings.input_tag).attr('class')
    const inputErrorField = element.closest('.' + (errorFieldClass.replace(/ /g, '.')))
    const label = form.find("label[for='" + (element.attr('id')) + "']:not(.message)")
    const labelErrorField = label.closest('.' + errorFieldClass)

    if (inputErrorField[0]) {
      if (element.attr('type') != 'checkbox') {
        inputErrorField.find('#' + (element.attr('id'))).detach()
        inputErrorField.replaceWith(element)
        label.detach()
        labelErrorField.replaceWith(label)
      } else {
        const container = element.parents('label');

        container.detach();
        inputErrorField.replaceWith(container);
      }
    }
  }
}
```
Not sure if this should go in the docs somewhere.

**Comment by tagliala** [comment]
Hi,

thanks for being part of the CSV community.

I'm successfully using CSV (simple forms) with Bootstrap's custom checkbox, without any additional rule.

I also suggest to place the input element inside the label tag, maybe you can take a look at bootstrap's implementation to refactor your template and retry without this extra rule

> Not sure if this should go in the docs somewhere.

We should document something at a higher level, because this does not depend on CSV but on developer's form implementation. Something like "By default, CSV tries to append the error tag to the closest $x element, if your template may have a different logic, you should customize `add` and `remove` javascript function"

Hope it helps

Closing here
